// botframework-webchat-component/lib/Activity/SendStatus.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.connectSendStatus = exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _botframeworkWebchatCore = require("botframework-webchat-core");

var _connectToWebChat = _interopRequireDefault(require("../connectToWebChat"));

var _Localize = _interopRequireWildcard(require("../Localization/Localize"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _Constants$ActivityCl = _botframeworkWebchatCore.Constants.ActivityClientState,
    SEND_FAILED = _Constants$ActivityCl.SEND_FAILED,
    SENDING = _Constants$ActivityCl.SENDING; // TODO: [P4] Currently, this is the only place which use a templated string
//       We could refactor this into a general component if there are more templated strings

function sendFailed(language, replace) {
  var text = (0, _Localize.localize)('SEND_FAILED_KEY', language);
  var retry = (0, _Localize.localize)('Retry', language);
  var match = /\{Retry\}/.exec(text);

  if (match) {
    return _react.default.createElement(_react.default.Fragment, null, text.substr(0, match.index), replace(retry), text.substr(match.index + match[0].length));
  } else {
    return text;
  }
}

var connectSendStatus = function connectSendStatus() {
  for (var _len = arguments.length, selectors = new Array(_len), _key = 0; _key < _len; _key++) {
    selectors[_key] = arguments[_key];
  }

  return _connectToWebChat.default.apply(void 0, [function (_ref, _ref2) {
    var focusSendBox = _ref.focusSendBox,
        language = _ref.language,
        postActivity = _ref.postActivity;
    var activity = _ref2.activity;
    return {
      language: language,
      retrySend: function retrySend(evt) {
        evt.preventDefault();
        postActivity(activity); // After clicking on "retry", the button will be gone and focus will be lost (back to document.body)
        // We want to make sure the user stay inside Web Chat

        focusSendBox();
      }
    };
  }].concat(selectors));
};

exports.connectSendStatus = connectSendStatus;

var _default = connectSendStatus(function (_ref3) {
  var styleSet = _ref3.styleSet;
  return {
    styleSet: styleSet
  };
})(function (_ref4) {
  var _ref4$activity$channe = _ref4.activity.channelData;
  _ref4$activity$channe = _ref4$activity$channe === void 0 ? {} : _ref4$activity$channe;
  var state = _ref4$activity$channe.state,
      language = _ref4.language,
      retrySend = _ref4.retrySend,
      styleSet = _ref4.styleSet;
  return _react.default.createElement("span", {
    "aria-live": "polite",
    className: styleSet.sendStatus
  }, state === SENDING ? _react.default.createElement(_Localize.default, {
    text: "Sending"
  }) : state === SEND_FAILED ? sendFailed(language, function (retry) {
    return _react.default.createElement("button", {
      onClick: retrySend,
      type: "button"
    }, retry);
  }) : false);
});

exports.default = _default;
