// react-dictate-button/lib/DictateButton.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _Composer = require('./Composer');

var _Composer2 = _interopRequireDefault(_Composer);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var DictateButton = function (_React$Component) {
  _inherits(DictateButton, _React$Component);

  function DictateButton(props) {
    _classCallCheck(this, DictateButton);

    var _this = _possibleConstructorReturn(this, (DictateButton.__proto__ || Object.getPrototypeOf(DictateButton)).call(this, props));

    _this.handleClick = _this.handleClick.bind(_this);
    _this.handleDictate = _this.handleDictate.bind(_this);
    _this.handleError = _this.handleError.bind(_this);

    _this.state = {
      started: false
    };
    return _this;
  }

  _createClass(DictateButton, [{
    key: 'handleClick',
    value: function handleClick(event) {
      this.props.onClick && this.props.onClick(event);
      !event.isDefaultPrevented() && this.setState(function (_ref) {
        var started = _ref.started;
        return { started: !started };
      });
    }
  }, {
    key: 'handleDictate',
    value: function handleDictate(event) {
      this.setState(function () {
        return { started: false };
      });
      this.props.onDictate && this.props.onDictate(event);
    }
  }, {
    key: 'handleError',
    value: function handleError(event) {
      this.setState(function () {
        return { started: false };
      });
      this.props.onError && this.props.onError(event);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var props = this.props,
          state = this.state;


      return _react2.default.createElement(
        _Composer2.default,
        {
          extra: props.extra,
          grammar: props.grammar,
          lang: props.lang,
          onDictate: this.handleDictate,
          onError: this.handleError,
          onProgress: props.onProgress,
          onRawEvent: props.onRawEvent,
          speechGrammarList: props.speechGrammarList,
          speechRecognition: props.speechRecognition,
          started: state.started && !props.disabled
        },
        function (context) {
          return _react2.default.createElement(
            'button',
            {
              className: props.className,
              disabled: context.readyState === 1 || context.readyState === 3 || !context.supported || props.disabled,
              onClick: _this2.handleClick
            },
            typeof props.children === 'function' ? props.children({ readyState: context.readyState }) : props.children
          );
        }
      );
    }
  }]);

  return DictateButton;
}(_react2.default.Component);

exports.default = DictateButton;


DictateButton.propTypes = {
  className: _propTypes2.default.string,
  disabled: _propTypes2.default.bool,
  extra: _propTypes2.default.any,
  grammar: _propTypes2.default.string,
  lang: _propTypes2.default.string,
  onClick: _propTypes2.default.func,
  onDictate: _propTypes2.default.func,
  onError: _propTypes2.default.func,
  onProgress: _propTypes2.default.func,
  onRawEvent: _propTypes2.default.func,
  speechGrammarList: _propTypes2.default.any,
  speechRecognition: _propTypes2.default.any
};
