// botframework-webchat-component/lib/Attachment/VimeoContent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _connectToWebChat = _interopRequireDefault(require("../connectToWebChat"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = (0, _connectToWebChat.default)(function (_ref) {
  var styleSet = _ref.styleSet;
  return {
    styleSet: styleSet
  };
})(function (_ref2) {
  var alt = _ref2.alt,
      autoPlay = _ref2.autoPlay,
      embedID = _ref2.embedID,
      loop = _ref2.loop,
      styleSet = _ref2.styleSet;
  var search = new URLSearchParams({
    autoplay: autoPlay ? 1 : 0,
    badge: 0,
    byline: 0,
    loop: loop ? 1 : 0,
    portrait: 0,
    title: 0
  }).toString();
  return (// TODO: We should encodeURI the URL
    _react.default.createElement("iframe", {
      "aria-label": alt,
      className: styleSet.vimeoContent,
      src: "https://player.vimeo.com/video/".concat(embedID, "?").concat(search)
    })
  );
});

exports.default = _default;
