// botframework-webchat-component/lib/Localization/da-DK.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function botSaidSomething(avatarInitials, text, timestamp) {
  return "Bot ".concat(avatarInitials, " sagde, ").concat(text, ", ").concat(xMinutesAgo(timestamp));
}

function userSaidSomething(avatarInitials, text, timestamp) {
  return "Bruger ".concat(avatarInitials, " sagde, ").concat(text, ", ").concat(xMinutesAgo(timestamp));
}

function xMinutesAgo(dateStr) {
  var date = new Date(dateStr);
  var dateTime = date.getTime();

  if (isNaN(dateTime)) {
    return dateStr;
  }

  var now = Date.now();
  var deltaInMs = now - dateTime;
  var deltaInMinutes = Math.floor(deltaInMs / 60000);
  var deltaInHours = Math.floor(deltaInMs / 3600000);

  if (deltaInMinutes < 1) {
    return 'Lige nu';
  } else if (deltaInMinutes === 1) {
    return 'Et minut siden';
  } else if (deltaInHours < 1) {
    return "".concat(deltaInMinutes, " minutter siden");
  } else if (deltaInHours === 1) {
    return "En time siden";
  } else if (deltaInHours < 5) {
    return "".concat(deltaInHours, " timer siden");
  } else if (deltaInHours <= 24) {
    return "Idag";
  } else if (deltaInHours <= 48) {
    return "Ig\xE5r";
  } else if (window.Intl) {
    return new Intl.DateTimeFormat('da-DK').format(date);
  } else {
    return date.toLocaleString('da-DK', {
      day: '2-digit',
      hour: '2-digit',
      hour12: false,
      minute: '2-digit',
      month: '2-digit',
      year: 'numeric'
    });
  }
}

var _default = {
  FAILED_CONNECTION_NOTIFICATION: 'Kunne ikke tilslutte',
  // Do not localize {Retry}; it is a placeholder for "Retry". English translation should be, "Send failed. Retry."
  SEND_FAILED_KEY: 'ikke sendt, {Retry}.',
  SLOW_CONNECTION_NOTIFICATION: 'Det tager længere tid at tilslutte end forventet',
  'Bot said something': botSaidSomething,
  'User said something': userSaidSomething,
  'X minutes ago': xMinutesAgo,
  // '[File of type '%1']': '[File of type '%1']",
  // '[Unknown Card '%1']': '[Unknown Card '%1']',
  'Adaptive Card parse error': 'Adaptive Card parse fejl',
  'Adaptive Card render error': 'Adaptive Card renderings-fejl',
  'Chat': 'Chat',
  'Download file': 'Hent fil',
  'Microphone off': 'Mikrofon slukket',
  'Microphone on': 'Mikrofon tændt',
  'Listening…': 'Lytter…',
  'Left': 'Venstre',
  'New messages': 'Ny besked',
  'Right': 'Højre',
  'retry': 'prøv igen',
  'Retry': '{retry}',
  // Please alter this value if 'Retry' at the beginning of a sentence is written differently than at the end of a sentence.
  'Sending': 'Sender',
  'Starting…': 'Starter...',
  'Tax': 'Skat',
  'Total': 'Total',
  'VAT': 'Moms',
  'Send': 'Send',
  'Speak': 'Tal',
  'Upload file': 'Upload fil',
  'Type your message': 'Skriv din besked'
};
exports.default = _default;
