// botframework-webchat-component/lib/Localization/de-DE.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function xMinutesAgo(dateStr) {
  var date = new Date(dateStr);
  var dateTime = date.getTime();

  if (isNaN(dateTime)) {
    return dateStr;
  }

  var now = Date.now();
  var deltaInMs = now - dateTime;
  var deltaInMinutes = Math.floor(deltaInMs / 60000);
  var deltaInHours = Math.floor(deltaInMs / 3600000);

  if (deltaInMinutes < 1) {
    return "jetzt";
  } else if (deltaInMinutes === 1) {
    return 'Vor einer Minute';
  } else if (deltaInHours < 1) {
    return "Vor ".concat(deltaInMinutes, " Minuten");
  } else if (deltaInHours === 1) {
    return "Vor eine Stunde";
  } else if (deltaInHours < 5) {
    return "Vor ".concat(deltaInHours, " Stunden");
  } else if (deltaInHours <= 24) {
    return "heute";
  } else if (deltaInHours <= 48) {
    return "gestern";
  } else if (window.Intl) {
    return new Intl.DateTimeFormat('de-DE').format(date);
  } else {
    return date.toLocaleString('de-DE', {
      day: '2-digit',
      hour: '2-digit',
      hour12: false,
      minute: '2-digit',
      month: '2-digit',
      year: 'numeric'
    });
  }
}

var _default = {
  FAILED_CONNECTION_NOTIFICATION: 'Keine Verbindung',
  // Do not localize {Retry}; it is a placeholder for "Retry". English translation should be, "Send failed. Retry."
  SEND_FAILED_KEY: 'konnte nicht senden, {Retry}.',
  SLOW_CONNECTION_NOTIFICATION: 'Eingeschränkte Konnektivität',
  'Chat': 'Chat',
  'Download file': 'Datei herunterladen',
  'Microphone off': 'Mikrofon aus',
  'Microphone on': 'Mikrofon aus',
  'Left': 'Links',
  'Listening…': 'hört zu…',
  'New messages': 'Neue Nachrichten',
  'retry': 'wiederholen',
  'Right': 'Rechts',
  'Retry': 'Wiederholen',
  // Please alter this value if 'Retry' at the beginning of a sentence is written differently than at the end of a sentence.
  'Send': 'Senden',
  'Sending': 'sendet…',
  'Speak': 'Sprechen',
  'Starting…': 'Startet…',
  'Tax': 'Steuer',
  'Total': 'Total',
  'Type your message': 'Verfasse eine Nachricht',
  'Upload file': 'Datei hochladen',
  'VAT': 'MwSt',
  'X minutes ago': xMinutesAgo
};
exports.default = _default;
