// botframework-webchat-component/lib/Localization/fr-FR.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function xMinutesAgo(dateStr) {
  var date = new Date(dateStr);
  var dateTime = date.getTime();

  if (isNaN(dateTime)) {
    return dateStr;
  }

  var now = Date.now();
  var deltaInMs = now - dateTime;
  var deltaInMinutes = Math.floor(deltaInMs / 60000);
  var deltaInHours = Math.floor(deltaInMs / 3600000);

  if (deltaInMinutes < 1) {
    return "À l'instant";
  } else if (deltaInMinutes === 1) {
    return 'Il y a une minute';
  } else if (deltaInHours < 1) {
    return "Il y a ".concat(deltaInMinutes, " minutes");
  } else if (deltaInHours === 1) {
    return "Il y a une heure";
  } else if (deltaInHours < 5) {
    return "Il y a ".concat(deltaInHours, " heures");
  } else if (deltaInHours <= 24) {
    return "Aujourd'hui";
  } else if (deltaInHours <= 48) {
    return "Hier";
  } else if (window.Intl) {
    return new Intl.DateTimeFormat('fr-FR').format(date);
  } else {
    return date.toLocaleString('fr-FR', {
      day: '2-digit',
      hour: '2-digit',
      hour12: false,
      minute: '2-digit',
      month: '2-digit',
      year: 'numeric'
    });
  }
}

var _default = {
  // FAILED_CONNECTION_NOTIFICATION: '',
  // Do not localize {Retry}; it is a placeholder for "Retry". English translation should be, "Send failed. Retry."
  SEND_FAILED_KEY: 'Échec d\'envoi, {Retry}.',
  // SLOW_CONNECTION_NOTIFICATION: '',
  'Chat': 'Discuter',
  // 'Download file': '',
  // 'Microphone off': '',
  // 'Microphone on': '',
  'Left': 'Gauche',
  'Listening…': 'Écoute…',
  'New messages': 'Nouveaux messages',
  'retry': 'Réessayer',
  'Right': 'Droite',
  'Retry': '{retry}',
  // Please alter this value if 'Retry' at the beginning of a sentence is written differently than at the end of a sentence.
  'Send': 'Envoyer',
  'Sending': 'Envoi…',
  'Speak': 'Parlez',
  'Starting…': 'Démarrage…',
  'Tax': 'Taxe',
  'Total': 'Total',
  'Type your message': 'Saisissez votre message',
  'Upload file': 'Envoyer fichier',
  'VAT': 'TVA',
  'X minutes ago': xMinutesAgo
};
exports.default = _default;
