// botframework-webchat-component/lib/Localization/nl-NL.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function botSaidSomething(avatarInitials, text, timestamp) {
  return "Bot ".concat(avatarInitials, " zei; ").concat(text, ", ").concat(xMinutesAgo(timestamp));
}

function userSaidSomething(avatarInitials, text, timestamp) {
  return "Gebruiker ".concat(avatarInitials, " zei; ").concat(text, ", ").concat(xMinutesAgo(timestamp));
}

function xMinutesAgo(dateStr) {
  var date = new Date(dateStr);
  var dateTime = date.getTime();

  if (isNaN(dateTime)) {
    return dateStr;
  }

  var now = Date.now();
  var deltaInMs = now - dateTime;
  var deltaInMinutes = Math.floor(deltaInMs / 60000);
  var deltaInHours = Math.floor(deltaInMs / 3600000);

  if (deltaInMinutes < 1) {
    return 'Zojuist';
  } else if (deltaInMinutes === 1) {
    return 'Een minuut geleden';
  } else if (deltaInHours < 1) {
    return "".concat(deltaInMinutes, " minuten geleden");
  } else if (deltaInHours === 1) {
    return "Een uur geleden";
  } else if (deltaInHours < 5) {
    return "".concat(deltaInHours, " uur geleden");
  } else if (deltaInHours <= 24) {
    return "Vandaag";
  } else if (deltaInHours <= 48) {
    return "Gisteren";
  } else if (window.Intl) {
    return new Intl.DateTimeFormat('nl-NL').format(date);
  } else {
    return date.toLocaleString('nl-NL', {
      day: '2-digit',
      hour: '2-digit',
      hour12: false,
      minute: '2-digit',
      month: '2-digit',
      year: 'numeric'
    });
  }
}

var _default = {
  FAILED_CONNECTION_NOTIFICATION: 'Verbinding maken niet mogelijk.',
  SEND_FAILED_KEY: 'Versturen mislukt, {retry}.',
  SLOW_CONNECTION_NOTIFICATION: 'Verbinding maken duurt langer dan normaal…',
  'Chat': 'Chat',
  'Download file': 'Bestand downloaden',
  'Microphone off': 'Microfoon uit',
  'Microphone on': 'Microfoon aan',
  'Listening…': 'Aan het luisteren…',
  'retry': 'probeer opnieuw',
  'Retry': 'Opnieuw proberen',
  'Send': 'Verstuur',
  'Sending': 'versturen',
  'Speak': 'Spreek',
  'Starting…': 'Starten…',
  'Tax': 'BTW',
  'Total': 'Totaal',
  'Type your message': 'Typ je bericht',
  'Upload file': 'Bestand uploaden',
  'VAT': 'VAT',
  'X minutes ago': xMinutesAgo
};
exports.default = _default;
