// botframework-webchat-component/lib/Localization/sv-SE.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function botSaidSomething(avatarInitials, text, timestamp) {
  return "Bot ".concat(avatarInitials, " sa, ").concat(text, ", ").concat(xMinutesAgo(timestamp));
}

function userSaidSomething(avatarInitials, text, timestamp) {
  return "Anv\xE4ndare ".concat(avatarInitials, " sa, ").concat(text, ", ").concat(xMinutesAgo(timestamp));
}

function xMinutesAgo(dateStr) {
  var date = new Date(dateStr);
  var dateTime = date.getTime();

  if (isNaN(dateTime)) {
    return dateStr;
  }

  var now = Date.now();
  var deltaInMs = now - dateTime;
  var deltaInMinutes = Math.floor(deltaInMs / 60000);
  var deltaInHours = Math.floor(deltaInMs / 3600000);

  if (deltaInMinutes < 1) {
    return 'Alldeles nyss';
  } else if (deltaInMinutes === 1) {
    return 'För en minut sen';
  } else if (deltaInHours < 1) {
    return "".concat(deltaInMinutes, " minuter sedan");
  } else if (deltaInHours === 1) {
    return "En timme sen";
  } else if (deltaInHours < 5) {
    return "".concat(deltaInHours, " timmar sen");
  } else if (deltaInHours <= 24) {
    return "Idag";
  } else if (deltaInHours <= 48) {
    return "Ig\xE5r";
  } else if (window.Intl) {
    return new Intl.DateTimeFormat('sv-SE').format(date);
  } else {
    return date.toLocaleString('sv-SE', {
      day: '2-digit',
      hour: '2-digit',
      hour12: false,
      minute: '2-digit',
      month: '2-digit',
      year: 'numeric'
    });
  }
}

var _default = {
  FAILED_CONNECTION_NOTIFICATION: 'Kunde inte ansluta.',
  // Do not localize {Retry}; it is a placeholder for "Retry". English translation should be, "Send failed. Retry."
  SEND_FAILED_KEY: 'kunde inte skicka, {Retry}.',
  SLOW_CONNECTION_NOTIFICATION: 'Det tar längre än vanligt att ansluta.',
  'Bot said something': botSaidSomething,
  'User said something': userSaidSomething,
  'X minutes ago': xMinutesAgo,
  // '[File of type '%1']': '[File of type '%1']",
  // '[Unknown Card '%1']': '[Unknown Card '%1']',
  'Adaptive Card parse error': 'Adaptive Card parse error',
  'Adaptive Card render error': 'Adaptive Card render error',
  'Chat': 'Chatt',
  'Download file': 'Ladda ned fil',
  'Microphone off': 'Mikrofon av',
  'Microphone on': 'Mikrofon på',
  'Left': 'Vänster',
  'Listening…': 'Lyssnar…',
  'New messages': 'Nya meddelanden',
  'Right': 'Höger',
  'retry': 'försök igen',
  'Retry': '{retry}',
  // Please alter this value if 'Retry' at the beginning of a sentence is written differently than at the end of a sentence.
  'Send': 'Skicka',
  'Sending': 'Skickar',
  'Speak': 'Läs upp',
  'Starting…': 'Startar…',
  'Tax': 'Skatt',
  'Total': 'Totalt',
  'Type your message': 'Skriv ditt meddelande',
  'Upload file': 'Ladd upp fil',
  'VAT': 'Moms'
};
exports.default = _default;
