// botframework-webchat-component/lib/Localization/zh-HANT.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function xMinutesAgo(dateStr) {
  var date = new Date(dateStr);
  var dateTime = date.getTime();

  if (isNaN(dateTime)) {
    return dateStr;
  }

  var now = Date.now();
  var deltaInMs = now - dateTime;
  var deltaInMinutes = Math.floor(deltaInMs / 60000);
  var deltaInHours = Math.floor(deltaInMs / 3600000);

  if (deltaInMinutes < 1) {
    return '剛剛';
  } else if (deltaInMinutes === 1) {
    return '一分鐘前';
  } else if (deltaInHours < 1) {
    return "".concat(deltaInMinutes, " \u5206\u9418\u524D");
  } else if (deltaInHours === 1) {
    return "\u4E00\u500B\u9418\u524D";
  } else if (deltaInHours < 5) {
    return "".concat(deltaInHours, " \u500B\u9418\u524D");
  } else if (deltaInHours <= 24) {
    return "\u4ECA\u65E5";
  } else if (deltaInHours <= 48) {
    return "\u6628\u65E5";
  } else if (window.Intl) {
    return new Intl.DateTimeFormat('zh-HANT').format(date);
  } else {
    return date.toLocaleString('zh-HANT', {
      day: '2-digit',
      hour: '2-digit',
      hour12: false,
      minute: '2-digit',
      month: '2-digit',
      year: 'numeric'
    });
  }
}

var _default = {
  FAILED_CONNECTION_NOTIFICATION: '接駁失敗。',
  // Do not localize {Retry}; it is a placeholder for "Retry". English translation should be, "Send failed. Retry."
  SEND_FAILED_KEY: '無法發送。{Retry}',
  SLOW_CONNECTION_NOTIFICATION: '接駁時間比平時長。',
  'Chat': '聊天',
  'Download file': '下載檔案',
  'Microphone off': '關掉麥克風',
  'Microphone on': '開啟麥克風',
  'Left': '左',
  'Listening…': '正在聆聽…',
  'New messages': '新訊息',
  'retry': '重試',
  'Retry': '{retry}',
  // Please alter this value if 'Retry' at the beginning of a sentence is written differently than at the end of a sentence.
  'Right': '右',
  'Send': '發送',
  'Sending': '正在發送',
  'Speak': '發言',
  'Starting…': '開始中…',
  'Tax': '稅',
  'Total': '總共',
  'Type your message': '請輸入您的訊息',
  'Upload file': '上載檔案',
  'VAT': '消費稅',
  'X minutes ago': xMinutesAgo
};
exports.default = _default;
