// botframework-webchat-component/lib/Localization/zh-HANS.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function xMinutesAgo(dateStr) {
  var date = new Date(dateStr);
  var dateTime = date.getTime();

  if (isNaN(dateTime)) {
    return dateStr;
  }

  var now = Date.now();
  var deltaInMs = now - dateTime;
  var deltaInMinutes = Math.floor(deltaInMs / 60000);
  var deltaInHours = Math.floor(deltaInMs / 3600000);

  if (deltaInMinutes < 1) {
    return '刚刚';
  } else if (deltaInMinutes === 1) {
    return '一分钟前';
  } else if (deltaInHours < 1) {
    return "".concat(deltaInMinutes, " \u5206\u949F\u524D");
  } else if (deltaInHours === 1) {
    return "\u4E00\u4E2A\u949F\u524D";
  } else if (deltaInHours < 5) {
    return "".concat(deltaInHours, " \u4E2A\u949F\u524D");
  } else if (deltaInHours <= 24) {
    return "\u4ECA\u65E5";
  } else if (deltaInHours <= 48) {
    return "\u6628\u65E5";
  } else if (window.Intl) {
    return new Intl.DateTimeFormat('zh-HANS').format(date);
  } else {
    return date.toLocaleString('zh-HANS', {
      day: '2-digit',
      hour: '2-digit',
      hour12: false,
      minute: '2-digit',
      month: '2-digit',
      year: 'numeric'
    });
  }
}

var _default = {
  FAILED_CONNECTION_NOTIFICATION: '接驳失败。',
  // Do not localize {Retry}; it is a placeholder for "Retry". English translation should be, "Send failed. Retry."
  SEND_FAILED_KEY: '无法发送。{Retry}',
  SLOW_CONNECTION_NOTIFICATION: '接驳时间比平时长。',
  'Chat': '聊天',
  'Download file': '下载文件',
  'Microphone off': '关掉麦克风',
  'Microphone on': '打开麦克风',
  'Left': '左',
  'Listening…': '正在倾听…',
  'New messages': '新讯息',
  'retry': '重试',
  'Retry': '{retry}',
  // Please alter this value if 'Retry' at the beginning of a sentence is written differently than at the end of a sentence.
  'Right': '右',
  'Send': '发送',
  'Sending': '正在发送',
  'Speak': '发言',
  'Starting…': '开始中…',
  'Tax': '税',
  'Total': '共计',
  'Type your message': '输入你的消息',
  'Upload file': '上传文件',
  'VAT': '消费税',
  'X minutes ago': xMinutesAgo
};
exports.default = _default;
