// botframework-webchat-component/lib/Localization/zh-YUE.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function xMinutesAgo(dateStr) {
  var date = new Date(dateStr);
  var dateTime = date.getTime();

  if (isNaN(dateTime)) {
    return dateStr;
  }

  var now = Date.now();
  var deltaInMs = now - dateTime;
  var deltaInMinutes = Math.floor(deltaInMs / 60000);
  var deltaInHours = Math.floor(deltaInMs / 3600000);

  if (deltaInMinutes < 1) {
    return '啱啱';
  } else if (deltaInMinutes === 1) {
    return '一分鐘前';
  } else if (deltaInHours < 1) {
    return "".concat(deltaInMinutes, " \u5206\u9418\u524D");
  } else if (deltaInHours === 1) {
    return "\u4E00\u500B\u9418\u524D";
  } else if (deltaInHours < 5) {
    return "".concat(deltaInHours, " \u500B\u9418\u524D");
  } else if (deltaInHours <= 24) {
    return "\u4ECA\u65E5";
  } else if (deltaInHours <= 48) {
    // https://zh-yue.wikipedia.org/wiki/尋日
    return "\u5C0B\u65E5";
  } else if (window.Intl) {
    return new Intl.DateTimeFormat('zh-HK').format(date);
  } else {
    return date.toLocaleString('zh-HK', {
      day: '2-digit',
      hour: '2-digit',
      hour12: false,
      minute: '2-digit',
      month: '2-digit',
      year: 'numeric'
    });
  }
}

var _default = {
  FAILED_CONNECTION_NOTIFICATION: '接駁唔倒。',
  // Do not localize {Retry}; it is a placeholder for "Retry". English translation should be, "Send failed. Retry."
  SEND_FAILED_KEY: '傳送唔倒。{Retry}。',
  SLOW_CONNECTION_NOTIFICATION: '接駁嘅時間比平時長。',
  'Chat': '傾偈',
  'Download file': '下載檔案',
  'Microphone off': '閂咪',
  'Microphone on': '開咪',
  'Left': '左',
  'Listening…': '聽緊你講嘢…',
  'New messages': '新訊息',
  'retry': '再嚟一次',
  'Retry': '{retry}',
  // Please alter this value if 'Retry' at the beginning of a sentence is written differently than at the end of a sentence.
  'Right': '右',
  'Sending': '而家傳送緊',
  'Starting…': '開始緊…',
  'Tax': '稅',
  'Total': '總共',
  'VAT': '消費稅',
  'Send': '傳送',
  'Speak': '講嘢',
  'Upload file': '上載檔案',
  'Type your message': '請打你嘅訊息',
  'X minutes ago': xMinutesAgo
};
exports.default = _default;
