// botframework-webchat-component/lib/Styles/adaptiveCardHostConfig.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _adaptivecards = require("adaptivecards");

var _defaultStyleSetOptions = _interopRequireDefault(require("./defaultStyleSetOptions"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// https://docs.microsoft.com/en-us/adaptive-cards/rendering-cards/host-config
var _default = function _default() {
  var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : _defaultStyleSetOptions.default,
      accent = _ref.accent,
      bubbleTextColor = _ref.bubbleTextColor,
      cardEmphasisBackgroundColor = _ref.cardEmphasisBackgroundColor,
      primaryFont = _ref.primaryFont,
      subtle = _ref.subtle;

  return new _adaptivecards.HostConfig({
    containerStyles: {
      default: {
        foregroundColors: {
          default: {
            default: bubbleTextColor,
            subtle: subtle
          },
          accent: {
            default: accent,
            subtle: accent
          }
        }
      },
      emphasis: {
        backgroundColor: cardEmphasisBackgroundColor,
        default: {
          default: bubbleTextColor,
          subtle: subtle
        }
      }
    },
    supportsInteractivity: true,
    fontFamily: primaryFont,
    imageSizes: {
      small: 40,
      medium: 80,
      large: 160
    },
    actions: {
      actionAlignment: 'stretch',
      actionsOrientation: 'vertical',
      buttonSpacing: 8,
      maxActions: 100,
      showCard: {
        actionMode: 'inline',
        inlineTopMargin: 8
      },
      spacing: 'default'
    },
    adaptiveCard: {
      allowCustomStyle: false
    },
    imageSet: {
      imageSize: 'medium',
      maxImageHeight: 100
    },
    factSet: {
      title: {
        color: 'default',
        size: 'default',
        isSubtle: false,
        weight: 'bolder',
        wrap: true,
        maxWidth: 150
      },
      value: {
        color: 'default',
        size: 'default',
        isSubtle: false,
        weight: 'default',
        wrap: true
      },
      spacing: 8
    }
  });
};

exports.default = _default;
