// web-speech-cognitive-services/lib/SpeechServices/TextToSpeech/fetchSpeechData.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _buildSSML = _interopRequireDefault(require("./buildSSML"));

var DEFAULT_LANGUAGE = 'en-US';
var DEFAULT_VOICE = 'Microsoft Server Speech Text to Speech Voice (en-US, JessaRUS)';
var SYNTHESIS_URL_TEMPLATE = 'https://{region}.tts.speech.microsoft.com/cognitiveservices/v1';

function _default(_x) {
  return _ref2.apply(this, arguments);
}

function _ref2() {
  _ref2 = (0, _asyncToGenerator2.default)(
  /*#__PURE__*/
  _regenerator.default.mark(function _callee(_ref) {
    var authorizationToken, _ref$lang, lang, outputFormat, pitch, rate, region, text, _ref$voice, voice, volume, ssml, url, res;

    return _regenerator.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            authorizationToken = _ref.authorizationToken, _ref$lang = _ref.lang, lang = _ref$lang === void 0 ? DEFAULT_LANGUAGE : _ref$lang, outputFormat = _ref.outputFormat, pitch = _ref.pitch, rate = _ref.rate, region = _ref.region, text = _ref.text, _ref$voice = _ref.voice, voice = _ref$voice === void 0 ? DEFAULT_VOICE : _ref$voice, volume = _ref.volume;
            ssml = (0, _buildSSML.default)({
              lang: lang,
              pitch: pitch,
              rate: rate,
              text: text,
              voice: voice,
              volume: volume
            });
            url = SYNTHESIS_URL_TEMPLATE.replace(/\{region\}/, region);
            _context.next = 5;
            return fetch(url, {
              headers: {
                Authorization: "Bearer ".concat(authorizationToken),
                'Content-Type': 'application/ssml+xml',
                'X-Microsoft-OutputFormat': outputFormat
              },
              method: 'POST',
              body: ssml
            });

          case 5:
            res = _context.sent;

            if (!(res.status !== 200)) {
              _context.next = 8;
              break;
            }

            throw new Error("Failed to syntheis speech, server returned ".concat(res.status));

          case 8:
            return _context.abrupt("return", res.arrayBuffer());

          case 9:
          case "end":
            return _context.stop();
        }
      }
    }, _callee, this);
  }));
  return _ref2.apply(this, arguments);
}
