// botframework-webchat-core/lib/sagas/sendMessageToPostActivitySaga.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = _callee;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _effects = require("redux-saga/effects");

var _whileConnected = _interopRequireDefault(require("./effects/whileConnected"));

var _sendMessage = require("../actions/sendMessage");

var _postActivity = _interopRequireDefault(require("../actions/postActivity"));

var _marked =
/*#__PURE__*/
_regenerator["default"].mark(_callee),
    _marked2 =
/*#__PURE__*/
_regenerator["default"].mark(sendMessageToPostActivity),
    _marked3 =
/*#__PURE__*/
_regenerator["default"].mark(postActivityWithMessage);

function _callee() {
  return _regenerator["default"].wrap(function _callee$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          _context.next = 2;
          return (0, _whileConnected["default"])(sendMessageToPostActivity);

        case 2:
        case "end":
          return _context.stop();
      }
    }
  }, _marked);
}

function sendMessageToPostActivity() {
  return _regenerator["default"].wrap(function sendMessageToPostActivity$(_context2) {
    while (1) {
      switch (_context2.prev = _context2.next) {
        case 0:
          _context2.next = 2;
          return (0, _effects.takeEvery)(function (_ref) {
            var payload = _ref.payload,
                type = _ref.type;
            return type === _sendMessage.SEND_MESSAGE && payload.text;
          }, postActivityWithMessage);

        case 2:
        case "end":
          return _context2.stop();
      }
    }
  }, _marked2);
}

function postActivityWithMessage(_ref2) {
  var _ref2$payload, method, text;

  return _regenerator["default"].wrap(function postActivityWithMessage$(_context3) {
    while (1) {
      switch (_context3.prev = _context3.next) {
        case 0:
          _ref2$payload = _ref2.payload, method = _ref2$payload.method, text = _ref2$payload.text;
          _context3.next = 3;
          return (0, _effects.put)((0, _postActivity["default"])({
            text: text,
            textFormat: 'plain',
            type: 'message'
          }, method));

        case 3:
        case "end":
          return _context3.stop();
      }
    }
  }, _marked3);
}
