// botframework-webchat-core/lib/sagas/sendPostBackToPostActivitySaga.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = _callee;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _effects = require("redux-saga/effects");

var _sendPostBack = require("../actions/sendPostBack");

var _postActivity = _interopRequireDefault(require("../actions/postActivity"));

var _whileConnected = _interopRequireDefault(require("./effects/whileConnected"));

var _marked =
/*#__PURE__*/
_regenerator["default"].mark(_callee),
    _marked2 =
/*#__PURE__*/
_regenerator["default"].mark(sendPostBackToPostActivity),
    _marked3 =
/*#__PURE__*/
_regenerator["default"].mark(postActivityWithPostBack);

function _callee() {
  return _regenerator["default"].wrap(function _callee$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          _context.next = 2;
          return (0, _whileConnected["default"])(sendPostBackToPostActivity);

        case 2:
        case "end":
          return _context.stop();
      }
    }
  }, _marked);
}

function sendPostBackToPostActivity() {
  return _regenerator["default"].wrap(function sendPostBackToPostActivity$(_context2) {
    while (1) {
      switch (_context2.prev = _context2.next) {
        case 0:
          _context2.next = 2;
          return (0, _effects.takeEvery)(function (_ref) {
            var payload = _ref.payload,
                type = _ref.type;
            return type === _sendPostBack.SEND_POST_BACK && payload.value;
          }, postActivityWithPostBack);

        case 2:
        case "end":
          return _context2.stop();
      }
    }
  }, _marked2);
}

function postActivityWithPostBack(_ref2) {
  var value;
  return _regenerator["default"].wrap(function postActivityWithPostBack$(_context3) {
    while (1) {
      switch (_context3.prev = _context3.next) {
        case 0:
          value = _ref2.payload.value;
          _context3.next = 3;
          return (0, _effects.put)((0, _postActivity["default"])({
            channelData: {
              postBack: true
            },
            text: typeof value === 'string' ? value : undefined,
            type: 'message',
            value: typeof value !== 'string' ? value : undefined
          }));

        case 3:
        case "end":
          return _context3.stop();
      }
    }
  }, _marked3);
}
