// react-say/lib/createMockSpeechSynthesisPonyfill.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.jestFnWithPromises = jestFnWithPromises;
exports["default"] = createMockSpeechSynthesisPonyfill;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _createDeferred = _interopRequireDefault(require("./createDeferred"));

var _events = _interopRequireDefault(require("events"));

function jestFnWithPromises(implementation) {
  var fn = jest.fn(function () {
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    fn.deferreds[fn.deferreds.length - 1].resolve(args);
    fn.deferreds = [].concat((0, _toConsumableArray2["default"])(fn.deferreds), [(0, _createDeferred["default"])()]);
    fn.promises = fn.deferreds.map(function (_ref) {
      var promise = _ref.promise;
      return promise;
    });
    implementation && implementation.apply(void 0, args);
  });
  var firstDeferred = (0, _createDeferred["default"])();
  fn.deferreds = [firstDeferred];
  fn.promises = [firstDeferred.promise];
  return fn;
}

function createMockSpeechSynthesisPonyfill() {
  var SpeechSynthesisUtterance =
  /*#__PURE__*/
  function () {
    function SpeechSynthesisUtterance(text) {
      (0, _classCallCheck2["default"])(this, SpeechSynthesisUtterance);
      this._events = new _events["default"]();
      this.text = text;
    }

    (0, _createClass2["default"])(SpeechSynthesisUtterance, [{
      key: "addEventListener",
      value: function addEventListener(name, handler) {
        this._events.addListener(name, handler);
      }
    }, {
      key: "dispatchEvent",
      value: function dispatchEvent(event) {
        this._events.emit(event.type, event);
      }
    }, {
      key: "removeEventListener",
      value: function removeEventListener(name, handler) {
        this._events.removeListener(name, handler);
      }
    }]);
    return SpeechSynthesisUtterance;
  }();

  return {
    speechSynthesis: {
      cancel: jestFnWithPromises(),
      getVoices: function getVoices() {
        return [{
          name: 'Cantonese',
          voiceURI: 'http://localhost/voice/zh-YUE'
        }];
      },
      speak: jestFnWithPromises()
    },
    SpeechSynthesisUtterance: SpeechSynthesisUtterance
  };
}
